In [1]:
%load_ext tutormagic

# Recursive Functions
Definition: A function in which the body of the function calls itself either directly or indirectly.

What does this mean?

Implication: Executing the body of a `recursive` function may require applying that function again

Recursion doesn't appear only in computer science. It is also used in art, nature and mathematics.

<img src = 'recursion.jpg' width = 500/>

## Digit Sums
As a starting example, let's analyze the sum of the digits of a number. 

In number `2013`, the digit sums is `2 + 0 + 1 + 3` = `6`.

Why do we do this? Because of the interesting properties of digit sums:

1. If a number `a` is divisible by `9`, then `digit_sum(a)` is also divisible by `9` 
2. Summing up digits is useful for typo detection

Below we have a credit card. Credit card numbers are long, and when humans type credit card numbers, it is possible to have a typo! Thus, in credit cards exist such thing called `checksum digit`, a 16 digit that's computed from all other digits and is not part of the account number. If the `checksum digit` doesn't match the computation of all other digits, then the credit card number is typed wrong (it's an invalid credit card number).

Credit cards use the **Luhn algorithm**, which we'll implement after `digit_sum`. 

## Sum Digits Without a While Statement
We can easily write a code that sums digits with a `while` statement. However, without a `while` statement, it would require `recursion`. 

Here we define the `split` function that takes a positive integer `n` and splits it to 2 parts:
1. All but its last digit
2. Its last digit

In [2]:
def split(n):
    """ Split positive n into all but its last digit and its last digit"""
    # n // 10 is all but n's last digit
    # n % 10 is n's last digit 
    return n // 10, n % 10

In [3]:
all_but_last, last = split(2013)
all_but_last

201

In [4]:
last

3

To sum the digits of `n`, we write the function `sum_digits`.

1. If `n` is less than `10` (if `n` is a single digit), then just return `n`
2. Otherwise, `split` `n` to `all_but_last` and `last` 
    * Then return `sum_digits(all_but_last) + last`
    
For the case of `2013`, we'll have:
1. `all_but_last` = `201`
2. `last` = `3`

In this cycle, `sum_digits` will return `sum_digits(201)` + `3`

In [5]:
def sum_digits(n):
    """ Return the sum of the digits of positive integer n"""
    # if n is less than 10, then just return n
    if n < 10:
        return n
    else:
        # split n
        all_but_last, last = split(n)
        return sum_digits(all_but_last) + last

Above is a `recursive` function because we called `sum_digits` within `sum_digits` (in the last line above, Python calls `sum_digits(all_but_last)`.

## The Anatomy of a Recursive Function
The `def` statement header of a recursive function is similar to other functions. 

In [6]:
def sum_digits(n):

SyntaxError: unexpected EOF while parsing (<ipython-input-6-1007e5bb2168>, line 1)

Typically, a recursive function starts with a **conditional statement that checks for `base cases`**.
* `base cases` are the simple version of the problem we are trying to solve. 

In the case of `sum_digits`, the `base case` is the case where `n` has only 1 digit left.

In [ ]:
if n < 10:
    return n

Base cases are evaluated **without recursive calls**. In the case above, it is done by just returning `n`. 

For non-base cases (e.g. `n` still have multiple digits), we have `recursive cases` that are evaluated **with recursive calls**. In the case of `sum_digits`, the recursive call is as the following,

In [ ]:
all_but_last, last = split(n)
return sum_digits(all_but_last) + last

Notice above that we don't try to call `sum_digits` on `n`. Instead, we call `sum_digits` on a simpler problem than we have before. If previously we have `n = 2013`, now we have `all_but_last = 201`, which is simpler since `all_but_last` has fewer digits to sum, getting us closer to the base case. 

Let's try to use `sum_digits`!

In [ ]:
sum_digits(2013)